### ISSUE TITLE
Specials on homepage are not updated when they are changed with cache on

### ISSUE THREAD

**Issue by PaccoRobin** [PrestaShop/PrestaShop#35542]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

We are about to upgrade from 1.6 to 8.1. Right now running a test site (version 8.1) parallel to the production site (version 1.6). Both sites have cache enabled.

We have the specials block (ps_specials) active on our homepage. Each month some products have discounts. On the production site old specials are taken off and new specials are displayed. But on the test site it still shows old specials. If you click on "All specials" it shows the new specials. If you return to the homepage it displays the old specials again.

If we turn off cache on the test site you can see the new specials on the homepage. But if you reactivate cache it shows the old specials again.

![Cache settings](https://github.com/PrestaShop/PrestaShop/assets/30149569/867eb668-a3f1-42a2-b012-fc9e665cc841)

### Expected behavior

We expect behavior like on the old site. When a new special is valid it should show it on the homepage and if a special is no longer valid it should disappear.

### Steps to reproduce

1. Make sure cache is on.
2. Check homepage and special products.
3. Go to products and create a valid special for a product. For example, 10% discount for the month March.
4. Check homepage. No valid special. Click all specials and see the valid special.
5. Turn off caching and check homepage. Valid special should appear.
6. Turn on caching and check homepage. There is no valid special.

### PrestaShop version(s) where the bug happened

8.1.3

### PHP version(s) where the bug happened

8.1.27

### If your bug is related to a module, specify its name and its version

ps_specials v.1.0.2

**Comment by Misty-94** [comment]
Unfortunately this is not the first problem with cache :(
There is a similar problem with featured products when you check the option to display randomly the product : with cache off it works as intended, with cache on the products are not displayed randomly.

**Comment by Hlavtox** [comment]
I don't know why a cache like this even exists on things as dynamic as products. We can load 24 products in a category dynamically, fast, but we need to cache 8 products on homepage? 🤔 

Ping @kpodemski

**Comment by kpodemski** [comment]
@Hlavtox yes, I do believe the homepage requires cache ;)

**Comment by AureRita** [comment]
Hi @PaccoRobin 

Thank you for your report, Despite my effort, I can't reproduce your issue as you can see here : 

[recording(123).webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/201ef3bb-6e3f-496e-9799-9ddc6ba1bef2)

Moreover you speak about ps_special, but at the start of the video, I don't see the link between the issue so I replace it by a catalog price rules, say it if I do it well or not, 

Did I miss something ? 

Waiting for your feedback

**Comment by PaccoRobin** [comment]
@AureRita 
See my video. I left out website details, but practically every time I move the arrow to the top I do a refresh of the site.
Hopefully this shows my problem with caching!

https://github.com/PrestaShop/PrestaShop/assets/30149569/149a1a20-8a15-41bf-889c-99b18a9e2ae2

**Comment by kpodemski** [comment]
@PaccoRobin if you save the product, on top of saving the specific price, it should clear the cache of the `ps_specials` module, the solution could be to clear the cache of this module after adding or updating the specific price

**Comment by PaccoRobin** [comment]
@kpodemski 

The problem arises when you prepare the specials. If I have 4 specials for February and I make 4 new specials for March in February, the specials for February should disappear on March 1st and new specials should be visible.

That does not happen. The old specials are still there and only if you turn off caching it works like it should. If you click on the sentence "more specials" it shows the new specials. Once you go back to the homepage you see the old specials.

It is not expected behaviour.

**Comment by kpodemski** [comment]
@PaccoRobin 

> The problem arises when you prepare the specials

Yes, that's true. There is no built-in mechanism to clear the cache of the modules in the future. The solution could be to invalidate the cache daily instead of having it as long as some products are updated.

Since you know you will use future discounts, you may want to add a CRON task to clear this block's cache daily.

**Comment by PaccoRobin** [comment]
@kpodemski 

Why does it work like a charm with Prestashop 1.6? We never had this problem until we started building a 8.1.3 website.

**Comment by PaccoRobin** [comment]
> Since you know you will use future discounts, you may want to add a CRON task to clear this block's cache daily.

How to do that as a temporary fix?

**Comment by kpodemski** [comment]
@PaccoRobin because cache invalidation worked differently:

https://github.com/PrestaShop/blockspecials/blob/master/blockspecials.php#L253-L258

It's been creating a cache daily.

**Comment by PaccoRobin** [comment]
> @PaccoRobin because cache invalidation worked differently:
> 
> https://github.com/PrestaShop/blockspecials/blob/master/blockspecials.php#L253-L258
> 
> It's been creating a cache daily.

@kpodemski 

So what to put in the cronjob to clear and create a cache daily for specials?
Are other modules affected by this behavior?

**Comment by PaccoRobin** [comment]
@kpodemski 
Hi,
Do you think this gets solved? Or is there a way to manually clear the cache daily?

**Comment by web-burn** [comment]
The same problem: when i remove or add discount to some product, the special block cache doesn't clear. It worked fine in the PrestaShop 1.7. Why it still is not solved? it is big bug

**Comment by web-burn** [comment]
> The same problem: when i remove or add discount to some product, the special block cache doesn't clear. It worked fine in the PrestaShop 1.7. Why it still is not solved? it is big bug

I have found the reason of this case problem when the ps_special block was not updated after add, remove or update a specific price for a product. The hooks actionObjectSpecificPriceCoreDeleteAfter, actionObjectSpecificPriceCoreAddAfter, actionObjectSpecificPriceCoreUpdateAfter were renamed to actionObjectSpecificPriceDeleteAfter', actionObjectSpecificPriceAddAfter', actionObjectSpecificPriceUpdateAfter. It is need to open the file ps_specials.php and replace actionObjectSpecificPriceCoreDeleteAfter, actionObjectSpecificPriceCoreAddAfter, actionObjectSpecificPriceCoreUpdateAfter by actionObjectSpecificPriceDeleteAfter', actionObjectSpecificPriceAddAfter', actionObjectSpecificPriceUpdateAfter.

**Comment by boo-code** [comment]
This is fixed, by @Hlavtox's own PR, and can be closed. The reason it stayed open is that PrestaShop/ps_specials#26 left its "Fixed ticket?" field empty, so nothing ever linked back here.

**What fixed it.** ps_specials#26, "Remove caching system and bump version to 2.0.0", merged 2026-04-12 and released as **v2.0.0 on 2026-05-21**, deletes the module's caching outright. With no cache there is nothing to invalidate, so both halves of this thread go away: the immediate one (a special added or removed does not refresh the block) and the one @PaccoRobin actually reported (a special that becomes valid or expires by the passage of time, with no save event to hook on). `upgrade-2.0.0.php` unregisters the invalidation hooks accordingly.

**Measured** on a 9.2.x shop running ps_specials 2.0.0, with `PS_SMARTY_CACHE` on, reading the ids the block renders and never clearing any cache between requests:

```
baseline                                                    block = [1, 2]
insert a specific price for product 7                       block = [1, 2, 7]
move its window to 2026-08-01 .. 2026-08-31 (expired)       block = [1, 2]
move its window to 2026-09-01 .. 2026-09-30 (valid)         block = [1, 2, 7]
```

The last two are the reported scenario exactly: nothing was saved through the back office and no hook fired, only the dates moved relative to now, and the block followed on the next page load.

Worth noting why `PS_SMARTY_CACHE` no longer matters here. `config/smarty.config.inc.php:20` sets `$smarty->caching = Smarty::CACHING_OFF`, and the only thing that raises it is `Tools::enableCache()`, which `Module::fetch()` calls solely when a non-null `$cache_id` is passed. 2.0.0 calls `$this->fetch($this->templateFile)` with no cache id, so the block renders uncached whatever that setting says. Confirmed at runtime: `PS_SMARTY_CACHE = '1'` while `Context::getContext()->smarty->caching = 0`.

**@web-burn** — your diagnosis was right, and it had already been fixed separately. `ObjectModel::getFullyQualifiedName()` is `str_replace('\\', '', get_class($this))`, and `get_class(new SpecificPrice())` is `SpecificPrice`, not `SpecificPriceCore`, so the emitted hooks are `actionObjectSpecificPriceAddAfter` and friends — verified by running it rather than reading it. The module's `upgrade/upgrade-1.0.4.php` unregisters the three `...Core...` names, deletes the stray hook rows they created, and registers the correct ones. So on anything from 1.0.4 onward the rename is handled; on 2.0.0 it is moot.

**For anyone still on 8.x**: `ps_versions_compliancy` for 2.0.0 is `min 8.1.0`, so updating the module is enough — no core upgrade needed, and no cron job as suggested earlier in the thread.

One caveat from my own shop, in case it confuses someone: if the module files are updated without the module upgrade actually running, `ps_hook_module` keeps the old `actionObjectSpecificPriceCore*After` rows. They are harmless, since 2.0.0 defines no such methods, but they make it look as though the old wiring is still in place.
